sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java

```diff
@@ -28,6 +28,25 @@
  */
 @Evolving
 public interface SupportsDelete {
+
+  /**
+   * Checks whether it is possible to delete data from a data source table that matches filter
+   * expressions.
+   * <p>
+   * Rows should be deleted from the data source iff all of the filter expressions match.
+   * That is, the expressions must be interpreted as a set of filters that are ANDed together.
+   * <p>
+   * Spark will call this method to check if the delete is possible without significant effort.
```

Comment:
`without significant effort` looks like a misleading assumption here. Some Data source can implement this by themself with significant efforts due to its own reason.

Reply:
This comes from the documentation for `deleteWhere`:

> Implementations may reject a delete operation if the delete isn't possible without significant effort. For example, . . .

I think that phrasing is a bit more clear because it uses "implementations may reject", so what constitutes "significant effort" is determined by the implementation.

I think a clearer way to say it here is to refer to that standard: "Spark will call this method to check whether `deleteWhere` would reject the delete operation because it requires significant effort."

It would also help to have more context: this is for some sources to determine whether or not a metadata delete can be performed.

Reply:
> I think a clearer way to say it here is to refer to that standard: "Spark will call this method to check whether deleteWhere would reject the delete operation because it requires significant effort."

This sounds better as there is a standard between `canDeleteWhere` and `deleteWhere`.

So `canDeleteWhere` is a much light-weight approach to know `deleteWhere` will reject a delete operation without actually calling `deleteWhere`.

Reply:
I agree with you @dongjoon-hyun @rdblue. I'll update the comment.

Reply:
> So canDeleteWhere is a much light-weight approach to know deleteWhere will reject a delete operation without actually calling deleteWhere.

Yes, @viirya. It actually provides us with a way to check if a data source is going to reject the delete via `deleteWhere` at planning time instead of just getting an exception during execution. In the future, we can use this functionality to decide whether Spark should rewrite this delete and execute a distributed query or it can just pass a set of filters.

Consider an example of a partitioned Hive table. If we have a delete predicate like `part_col = '2020'`, we can just drop the matching partition to satisfy this delete. In this case, the data source should return `true` from `canDeleteWhere` and use the filters it accepts in `deleteWhere` to drop the partition. At the same time, if we have a delete predicate like `id = 10`, Hive tables would not be able to execute this delete using a metadata only operation without rewriting files. In that case, the data source should return `false` from `canDeleteWhere` and we should use a more sophisticated row-level API to find out which records should be removed (the API is yet to be discussed, but we need this PR as a basis).

If we decide to support subqueries and all delete use cases by simply extending the existing API, this will mean all data sources will have to implement a lot of Spark logic to determine which records changed. I don't think we want to go that way as the Spark logic to determine which records should be deleted is independent of the underlying data source. So the assumption is that Spark will execute a plan to find which records must be deleted for data sources that return `false` from `canDeleteWhere`.

Reply:
My worry is that we may change the API back and forth if we don't have a clear big picture. Right now this patch is not useful as it only changes where we throw the exception, but I can see that this will be useful when we have the row-level delete API and we can use the `canDeleteWhere` to decide if we want to use the row-level API or not.

Do we have a sketch about how the row-level API would look like? It seems weird if we merge a patch that is useful to something that is not being designed yet.

Reply:
Mind filing JIRAs for the cases you'd leverage this API? I would prefer to have an explicit plan for that since three committers asked the same question. It's also best to move such description into the PR description.

Reply:
I've updated the PR description with more information from my comment above.

The design doc is being prepared. It felt to me that the community has enough confidence that we will need such a check at planning time (maybe, I misunderstand concepts and expectations behind `SupportsDelete`). In my view, it was pretty independent of the design of the row-level API and that's why I submitted it now.

Reply:
I've updated the comment, @dongjoon-hyun @rdblue.

Reply:
I don't think simply planning time vs runtime justifies the API that dose the same thing that another API does. `SupportsDelete.deleteWhere` documents its purpose of failure cases which can fail fast as well (although it's runtime). 

Since it's planned to be used for something else as you elaborated, I think it should be best to make sure we all know the full context. We can think about other shapes of the API, for example, returning unhandled filters to make the error message to show which filters are not supported.

Reply:
If we want to return a set of rejected filters to form a better message, that is OK for me. I am also not going to push for this change if there is no consensus. As I wrote before, it felt that the community had enough confidence that we would eventually need such a check at planning time based on how `SupportsDelete` is designed. If we want to delay this conversation until later, fine with me too.

Reply:
> My worry is that we may change the API back and forth if we don't have a clear big picture. Right now this patch is not useful as it only changes where we throw the exception, but I can see that this will be useful when we have the row-level delete API and we can use the canDeleteWhere to decide if we want to use the row-level API or not.

This is exactly the reason for adding this API. It is a step toward rewriting plans for row-level DELETE and MERGE operations. The current `deleteWhere` exception approach happens while running the physical plan, when it is too late to rewrite the plan for row-level changes. Adding the `canDeleteWhere` check fixes that problem.

Since you can easily see how it will be used, what is the concern about adding this?

Reply:
My concern is not about this PR itself, but about the order of commits. It seems more natural to have the row-level delete API first, and then do this change. At that time we can have tests to verify if we can switch correctly.

If you are working with @aokolnychyi on this feature, and you two think this is better for your development, please go ahead and merge it.

Reply:
Thanks, we do think that it is helpful to have it in this order. This was an early problem that we ran into and the solution is clear.

Reply:
Can you guys just file related JIRAs for that before merging? That's all I asked.

Reply:
Just to update this thread, JIRAs were created. [Here](https://github.com/apache/spark/pull/30562#issuecomment-737779083) is a summary.